### ISSUE TITLE
Automatically inject default variables into template

### ISSUE THREAD

**Issue by JordanSussman** [go-vela/community#72]
## Description

[Templates](https://go-vela.github.io/docs/templates/overview/) should automatically have some of the default variables injected to reduce boilerplate variable injection that would otherwise be need.

### Current logic

template.yml

```yaml
metadata:
  template: true

{{$br := .branch}}

steps:

  - name: test
    commands:
      - go test ./...
    image: {{ .image }}
    {{ .pull_policy }}
    ruleset:
      event: [ push, pull_request ]

  # if branch equals master add this step to the final pipeline
  {{ if (eq $br "master") }}

  - name: build
    commands:
      - go build
    image: {{ .image }}
    {{ .pull_policy }}
    ruleset:
      event: [ push, pull_request ]

  {{ end }}
```

vela.yml

```yaml
version: "1"
templates:
  - name: sample
    source: github.com/<org>/<repo>/path/to/file/<template>.yml
    type: github

steps:
  - name: sample
    template:  
      name: golang
      vars:
        image: golang:latest
        pull_policy: "pull: true"
        branch: master
```

### Proposed logic

template.yml

```yaml
metadata:
  template: true

{{$br := getVelaBranch }}

steps:

  - name: test
    commands:
      - go test ./...
    image: {{ .image }}
    {{ .pull_policy }}
    ruleset:
      event: [ push, pull_request ]

  # if branch equals master add this step to the final pipeline
  {{ if (eq $br "master") }}

  - name: build
    commands:
      - go build
    image: {{ .image }}
    {{ .pull_policy }}
    ruleset:
      event: [ push, pull_request ]

  {{ end }}
```

vela.yml

```yaml
version: "1"
templates:
  - name: sample
    source: github.com/<org>/<repo>/path/to/file/<template>.yml
    type: github

steps:
  - name: sample
    template:  
      name: golang
      vars:
        image: golang:latest
        pull_policy: "pull: true"
```

Mixed opinions if this should utilize custom [functions](https://golang.org/pkg/text/template/#hdr-Functions) or if we merge the vars and system vars together.

## Value

- Reduce the boilerplate variable injection needed
- Templates can ensure that variable values are consistent/correct

## Definition of Done

Templates are able to determine information about the pipeline without needing to inject vars

## Effort  (Optional)

couple of days

## Impacted Personas (Optional)

Anyone that utilizes templates

**Comment by JordanSussman** [comment]
Looking for thoughts on utilizing a template function named `vela` that returns the variable value:

```go
package main

import (
	"os"
	"strings"
	"text/template"

	"github.com/Masterminds/sprig"
)

func main() {
	// create test map with combination of platform and user variables
	originalEnvs := map[string]string{
		"VELA_REPO": "foo",
		"VELA_USER": "bar",
		"FOO":       "test1",
		"BAR":       "test2",
	}

	// create another map to hold our modified envs
	newEnvs := make(map[string]string)
	for key, value := range originalEnvs {
		// ensure the key has the prefix of VELA
		if strings.HasPrefix(key, "VELA") {
			// remove the VELA_ prefix and lowercase the string
			newEnvs[strings.ToLower(strings.TrimPrefix(key, "VELA_"))] = value
		}
	}

	// create map that will contain our vela function
	templateFuncMap := map[string]interface{}{
		"vela": func(input string) string {
			// check if key exists within map
			if _, ok := newEnvs[input]; ok {
				// return value if exists
				return newEnvs[input]
			}
			// return empty string if not exists
			return ""
		},
	}

	tmpl := `
Exists {{ vela "repo" }}
Nonexist {{ vela "bad" }}
	`

	t, err := template.New("test").Funcs(sprig.TxtFuncMap()).Funcs(templateFuncMap).Parse(tmpl)
	if err != nil {
		panic(err)
	}

	err = t.Execute(os.Stdout, nil)

}
```

You can run the above code [here](https://play.golang.org/p/vKlXyAETbz7) to receive the output of:

```
Exists foo
Nonexist
```

TL;DR utilizing the syntax of `{{ vela "repo"}}` within the template would return the value of the `VELA_REPO` variable if it exists.

**Comment by jbrockopp** [comment]
I think this looks and sounds good 👍 

At first I had some concerns around determining variables with multiple `_` in them.

But it appears that isn't a problem based off my modified version of what you were testing:

https://play.golang.org/p/y9iqbqfJ-U1

Looks like all that is necessary is ensuring the proper field name is used after the `vela` function which seems fine:

```yaml
{{ vela "repo_full_name" }}
```

would be used to return the value of the `VELA_REPO_FULL_NAME` environment variable 👍 

 I'll bring this up with the team today to see if they have any further feedback.
